RedisLookupServer: Batch requests

orozery opened the pull request:
This PR transformers RedisLookupServer to be fully asynchronous.

The basic flow is that API functions such as insert, remove and lookup simply create a "WorkItem" in an async priority queue (where lookups have the highest priority).
We add a single python thread which reads from this queue and batches requests which are sent to the redis server.
For lookups, we wait for a future which is set when the lookup gets back it response from the batch.

There is still an optimization to consider for caching repeated lookups (there are actually 2 identical lookups triggered per each kv loading).

orozery committed:
RedisLookupServer: Batch requests

This commit changes the RedisLookupServer to be fully asynchronous,
and furthermore to batch requests.

Signed-off-by: Or Ozeri <oro@il.ibm.com>

vMaroon commented on lmcache/experimental/lookup_server/redis_server.py:
```diff
@@ -12,13 +15,29 @@
 
 logger = init_logger(__name__)
 
+BATCH_SIZE = 10
```
The tuning of these parameters should be benchmark driven, so it's best to make them configurable from env.

vMaroon replied:
While we're here, maybe now is the right time to split the configuration of sharing from indexing properly? (instead of the port hack). Not a PR blocker.

vMaroon commented on lmcache/experimental/lookup_server/redis_server.py:
```diff
@@ -28,31 +47,128 @@ def __init__(self, config: LMCacheEngineConfig):
+                    self._add_to_pipeline(pipe, item)
+                    items.append(item)
+                logger.debug(f"Sending a batch of {len(items)} requests")
+                results = pipe.execute()
```
Is this call guaranteed to not throw? As far as I know - it could. Better wrap this with a try.

orozery replied:
You're right. I pushed a fix.
For the rest of the comments I will add a configuration parameter (plan for tomorrow).

vMaroon commented on lmcache/experimental/lookup_server/redis_server.py:
```diff
@@ -28,31 +47,128 @@ def __init__(self, config: LMCacheEngineConfig):
         #decode_responses=False)
 
-    def _get_indexing_key(self, key: CacheEngineKey):
+        self.queue: asyncio.PriorityQueue[tuple[_WorkPriority, _WorkItem]] = (
```
What happens in cases of backpressure? Can this eat up memory up to an OOM?

vMaroon commented on lmcache/experimental/lookup_server/redis_server.py:
```diff
@@ -28,31 +47,128 @@ def __init__(self, config: LMCacheEngineConfig):
+        asyncio.run_coroutine_threadsafe(
+            self.queue.put((_WorkPriority.LOOKUP, item)), self.loop)
+
+        cfut = asyncio.run_coroutine_threadsafe(_wait_for_future(fut),
```
Can we configure the timeout here? If an answer does not come quickly enough - better not block the caller.

orozery committed:
Add lookup server configuration parameters

This commit adds parameters for configuring the lookup server.
Additionally, the lookup server can now be enabled without setting enable_p2p
(for indexing without sharing).

Signed-off-by: Or Ozeri <oro@il.ibm.com>

orozery commented:
@vMaroon Ready for another review
Note that now you can enable indexing without sharing simply by setting `lookup_url` and `distributed_url` with `enable_p2p=false`.
